### PR TITLE
feat(api): OpenAI sampling params + finish_reason + streaming usage (#14)

### LINKED ISSUE #14: Surface sampling parameters in /v1/chat/completions (top_p, top_k, seed, stop, finish_reason, logprobs)
## Summary

The `/v1/chat/completions` endpoint accepts only `model`, `messages`, `max_tokens`, `temperature`, `stream` (`crates/tahoma-api/src/lib.rs:197-206`). Industry-standard fields are missing: `top_p`, `top_k`, `seed`, `stop`, `frequency_penalty`, `presence_penalty`, `logprobs`, `top_logprobs`. Streaming responses lack `usage` deltas, and `finish_reason` is always `"stop"` regardless of actual stop reason (`tahoma-api/src/lib.rs:417`).

`sparse-moe` already has `SamplingConfig` internally with `top_p` / `top_k` / `rep_penalty` / window. This issue surfaces those + several others at the HTTP API boundary and propagates them to all engines.

## Why

Tahoma's API hands non-default OpenAI clients an incomplete experience. Many evals (lm-eval-harness, llm-foundry, simple-evals) and downstream tools (LangChain, LlamaIndex, dspy, OpenAI Agents SDK) set these fields and expect them honored. Today they're silently dropped — clients have no way to tell that `top_p=0.9` they sent had zero effect.

`finish_reason` differentiation matters for observability — distinguishing "model finished naturally" (`stop`), "hit max_tokens" (`length`), "EOS token emitted" (also `stop` per spec), "explicit stop string matched" (`stop`), "client cancelled" (custom or `null`) is signal for routing decisions, retries, and quality checks.

### (1) Extend `ChatCompletionRequest`

**File:** `crates/tahoma-api/src/lib.rs:197-206`

Add fields:
```rust
pub struct ChatCompletionRequest {
    pub model: String,
    pub messages: Vec<ChatMessage>,
    #[serde(default)]
    pub max_tokens: Option<u32>,
    #[serde(default)]
    pub temperature: Option<f32>,
    #[serde(default)]
    pub top_p: Option<f32>,
    #[serde(default)]
    pub top_k: Option<u32>,
    #[serde(default)]
    pub seed: Option<u64>,
    #[serde(default)]
    pub stop: Option<StopSpec>, // string or array of strings, OAI-style
    #[serde(default)]
    pub frequency_penalty: Option<f32>,
    #[serde(default)]
    pub presence_penalty: Option<f32>,
    #[serde(default)]
    pub logprobs: Option<bool>,
    #[serde(default)]
    pub top_logprobs: Option<u32>,
    #[serde(default)]
    pub stream: Option<bool>,
    #[serde(default)]
    pub stream_options: Option<StreamOptions>,
}

#[derive(Deserialize)]
#[serde(untagged)]
pub enum StopSpec {
    Single(String),
    Multiple(Vec<String>),
}

#[derive(Deserialize)]
pub struct StreamOptions {
    #[serde(default)]
    pub include_usage: bool,
}
```

### (2) Extend the internal sampling type

**File:** `crates/tahoma-engine-sparse-moe/src/sampling.rs`

`SamplingConfig` exists. Add `frequency_penalty: f32`, `presence_penalty: f32`, `seed: Option<u64>`, `stop: Vec<String>`. The first two apply during logit shaping (count token occurrences in the current generation, subtract penalty); `seed` initializes the RNG before sampling; `stop` is matched against the running detokenized text after each step.

For the multi-stage Forward frame: `SAMPLING_WIRE_BYTES` will need to grow to accommodate the new fields. **Bump the frame-kind version** (currently `0x53_4D_45_02` Forward) to `0x03` so old peers fail fast at `parse_kind` rather than mis-parsing. See `crates/tahoma-engine-sparse-moe/src/dist.rs:30` for the frame versioning convention.

### (3) Route sampling through every engine

| Engine | Path | Notes |
|---|---|---|
| `sparse-moe` | `crates/tahoma-engine-sparse-moe/src/sampling.rs` + `runner.rs` | Already has the structure |
| `ov-genai` | `crates/tahoma-engine-openvino/src/genai.rs` | OV-GenAI's `GenerationConfig` already supports all these — translate directly |
| `ov-runtime` | `crates/tahoma-engine-openvino/src/runtime.rs` | Logits applied client-side; implement sampler there |
| `ov-dist-spec` | `crates/tahoma-engine-openvino/src/dist_spec.rs` | Same as runtime |

### (4) `finish_reason` differentiation

**File:** `crates/tahoma-types/src/lib.rs` (or wherever `Chunk` is defined)

Add a `FinishReason` enum:
```rust
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FinishReason {
    Stop,        // EOS token or stop-string match
    Length,      // max_tokens hit
    Cancelled,   // client disconnect / explicit /v1/cancel
}
```

Have each engine emit the right one in the final chunk. API converts to OpenAI's string form (`stop`, `length`, `null` for in-progress streaming).

Current behavior: `tahoma-api/src/lib.rs:417` always emits `"stop"`. Change to read from `Chunk::finish_reason`.

### (5) Streaming usage deltas

OpenAI's SSE stream emits a final chunk with `usage: { prompt_tokens, completion_tokens, total_tokens }` when the client sets `stream_options.include_usage: true`. Implement same — Tahoma already has `prompt_tokens` and emits per-token chunks, so summation is trivial.

### (6) logprobs

OV-GenAI's `GenerationConfig` has a `logprobs` field; sparse-MoE samples logits locally and has the data on hand.

- `ChatCompletionRequest.logprobs` → engine generation config
- Engine emits `Chunk` with `logprobs: Some(Vec<TokenLogprob>)`
- API serializes per OpenAI schema: `{ "content": [ { "token": "...", "logprob": -1.23, "bytes": [...], "top_logprobs": [...] } ] }`

**Note:** `Chunk::logprobs` is currently `Option<()>` (a placeholder per the codebase audit). Define the real type.

Skip `logit_bias` for this issue — adds tokenizer complexity. Separate issue.

## Acceptance criteria

- [ ] `/v1/chat/completions` accepts all listed fields without 400 errors
- [ ] `top_p`, `top_k`, `seed`, `stop`, `frequency_penalty`, `presence_penalty` are honored by the `sparse-moe` engine. Smoke test: `stop: ["Berlin"]` truncates "Paris. The capital of Germany is Berlin" before "Berlin"
- [ ] Same sampling fields honored by `ov-genai` (passed through to `LLMPipeline.GenerationConfig`)
- [ ] `finish_reason` is `"length"` when `max_tokens` is hit, `"stop"` when EOS or stop-string is matched. Smoke test: `max_tokens: 5` returns `finish_reason: "length"`
- [ ] `logprobs: true` returns per-token logprobs in the response (OAI schema), at minimum for the `ov-genai` and `sparse-moe` engines
- [ ] `stream: true` with `stream_options: { include_usage: true }` emits a final SSE chunk containing `usage`
- [ ] Multi-stage pipeline-parallel respects the new sampling fields (verified by 2-box smoke test)
- [ ] Sampling-frame version bumped to `0x53_4D_45_03` (or whatever the next free low byte is); old peers reject the new frame with a clear error at `parse_kind`

## Effort

~ 2–3 days. logprobs alone is ~1 day; the rest is wiring + multi-stage frame bump + test coverage.

## Out of scope

- `logit_bias` — needs tokenizer integration; separate issue
- `tool_choice` / `tools` (function calling) — much larger work
- `response_format: json_schema` / grammar-constrained sampling — separate larger work
- `audio` / multimodal — out of scope

## References

- [OpenAI Chat Completions API reference](https://platform.openai.com/docs/api-reference/chat/create)
- [OpenVINO GenAI GenerationConfig](https://openvinotoolkit.github.io/openvino.genai/docs/structov_1_1genai_1_1GenerationConfig.html)
- Existing sparse-MoE SamplingConfig: `crates/tahoma-engine-sparse-moe/src/sampling.rs`
- Frame-kind versioning convention: `crates/tahoma-engine-sparse-moe/src/dist.rs:30`

### PR DESCRIPTION
**Pull request by t8**
Closes #14.

> **Stacks on #90 (#55).** This branch is cut from `fix/ovgenai-usage-tokens-55`, so it contains that commit too (`b4a58bc`). Merge #90 first, or this will be rebased onto main after it lands. The #14 work is the second commit (`21c15f2`).

Surfaces the standard OpenAI sampling + observability fields on `/v1/chat/completions` and routes them through the engines.

## What's in
| Area | Change |
|---|---|
| **Request** | `top_p`, `top_k`, `seed`, `stop` (string\|array), `frequency_penalty`, `presence_penalty`, `logprobs`, `top_logprobs`, `stream_options` — bundled into `cascadia_types::SamplingParams` on `GenerationTask`. |
| **finish_reason** | New `FinishReason` enum on the final `Chunk`; API emits `"length"` when the engine hit `max_tokens`, else `"stop"` (was hardcoded `"stop"`). |
| **Sampler (sparse-moe)** | `SamplingConfig` + `sample()` gain `top_k`, `frequency_penalty`, `presence_penalty`. |
| **ov-genai** | Forwards `top_p`/`top_k`/`seed`/penalties into `ov::genai::GenerationConfig` via new shim setters. |
| **Multi-stage wire** | `SAMPLING_WIRE_BYTES` 28→40; `Forward`/`ForwardBatch` codes bumped `0x02→0x04` / `0x03→0x05` so a peer on the old 28-byte layout fails fast at `parse_kind` (the issue's suggested `0x03` was already taken by `ForwardBatch`). |
| **Streaming** | `stream_options.include_usage` emits a final `usage` SSE chunk. |
| **stop** | Stop sequences truncate the assembled output at the earliest match (API-level). |
| **logprobs** | Request + OpenAI response schema (`ChatLogprobs`) wired; surfaced when chunks carry per-token logprobs. |

## Validation status (honest)
**Unit/integration tested (stub build, no hardware):**
- ✅ Sampler: top_k forces argmax under temperature; frequency penalty count-scales; presence penalty is count-independent.
- ✅ Wire: 40-byte round-trip incl. new fields; frame-code bump + old-code rejection.
- ✅ API: `finish_reason` length-vs-stop, stop-sequence truncation, full-param-set acceptance (no 400), streaming `usage` frame.
- ✅ `cargo fmt --all --check` clean; full workspace test green; no regressions.

**Needs on-HW validation before merge (no OpenVINO/Intel on this dev box):**
- ⏳ ov-genai actually honoring the new `GenerationConfig` knobs — the C++ shim setters are a mechanical mirror of the existing ones over standard OV fields, and the **AI-PC CI tier compiles the openvino path**, but live behavior (e.g. `seed` reproducibility, `top_p` effect) should be confirmed on a real model.
- ⏳ 2-box multi-stage sparse-moe carrying the new sampling fields end-to-end (wire round-trips in-process; a real 2-rank run confirms the bump).

## Deferred follow-ups (called out, not silently dropped)
- **logprobs emission per engine.** Request + response schema are in place, but no engine emits per-token logprobs yet: ov-genai needs `DecodedResults.scores` extraction through the shim, and sparse-moe single-stage emits one final chunk so it needs per-token emission (or a `Vec<TokenLogprobs>` on the final chunk). `logprobs: true` is accepted and returns `null` until then.
- **finish_reason for ov-runtime / ov-dist-spec / gemma4** (currently fall back to `"stop"`, unchanged).
- **Engine-side early-stop on `stop` strings** (would also save compute; today truncation is API-level output-only).
- **Streaming stop-sequence trimming** across chunk boundaries.

### CODE REVIEW

**Comment by t8** [comment]
Addressed the two review NITs in 5a5155b:
- **NIT-1 (corrupt zero `repetition_penalty`):** `decode_sampling` now clamps `<= 0`/NaN back to the 1.0 no-op (min = `MIN_POSITIVE`), so an out-of-version peer can't feed a 0 divisor into the sampler (`logit / α = +inf`). Added a regression test `decode_sanitizes_corrupt_zero_repetition_penalty`.
- **NIT-2 (penalty scope):** documented in-code that frequency/presence penalties deliberately span the full history (prompt + completion), matching the existing `repetition_penalty` behavior — a conscious divergence from OpenAI's completion-only scope.

The AI-PC CI job is the remaining gate (it compiles the C++ shim + exercises ov-genai); a reviewer separately verified all five new `GenerationConfig` setters map to real public members of the targeted OV 2026.2.0.0 header.

**Comment by t8** [comment]
## On-hardware validation (alpha Panther Lake + charlie Lunar Lake, real OpenVINO 2026.1)

Built `--features openvino` on the NUCs and exercised the live engines. Results:

**ov-genai (CPU, TinyLlama int4-ov):** 7/8 ✅ — `top_p`/`top_k`/penalties accepted (the C++ shim setters work on real OV, no crash), `finish_reason=length`, stop truncation, streaming `usage` frame all pass. The 1 fail ↓.
**ov-runtime (iGPU, Llama-3.2-1B shard):** ✅ — sampling accepted; **`seed` reproducible** (Rust sampler).
**sparse-moe 2-stage pipeline (loopback, m2_tiny):** ✅ — a request carrying the extended sampling block (`top_k`/`frequency_penalty`/`seed`) completes across the **bumped 40-byte Forward frame** end-to-end.

### Finding: ov-genai `seed` not reproducible across requests
Two `seed=42` requests on ov-genai → different output; on ov-runtime → identical. **Wiring is correct** — root cause is upstream: OV-GenAI's `StatefulLLMPipeline::generate` only re-seeds when the seed *value changes* between consecutive calls (`if (m_sampler.get_seed() != config.rng_seed) set_seed(...)`) and never clears the per-request RNG context, so repeated identical seeds reuse the advanced RNG. `seed` is still forwarded (different seeds change output), and the Rust-sampler engines honor it fully.

Documented at the wiring site in e525485. Full per-request reproducible ov-genai needs the `ContinuousBatchingPipeline` (fresh request context per request) — i.e. **#20**, not a #14-scope change.
